Add /account/{address}

dynco-nym opened the pull request:
- meant to replace `https://explorer.nymtech.net/api/v1/tmp/unstable/account/{address}` for Explorer v2

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nymtech/nym/5673)
<!-- Reviewable:end -->

dynco-nym committed:
Add /account/{address}

dynco-nym committed:
Don't query vesting info

dynco-nym committed:
Don't query rewards

dynco-nym committed:
Remove unused code

dynco-nym committed:
Fix clippy

jstuczyn commented on nym-api/build.rs:
```diff
@@ -1,13 +1,20 @@
 use sqlx::{Connection, FromRow, SqliteConnection};
-use std::env;
+use std::{env, fs::Permissions, os::unix::fs::PermissionsExt};
```
this fails windows build

dynco-nym replied:
see [other reply](https://github.com/nymtech/nym/pull/5673#discussion_r2044546718)

jstuczyn commented on nym-api/build.rs:
```diff
@@ -62,3 +69,26 @@ async fn main() {
+    let mut file = File::create("enter_db.sh").await?;
+    let contents = format!(
+        "#!/bin/bash\n\
+        sqlite3 -init settings.sql {}/{}",
```
can we guarantee `sqlite3` exists on every machine? why not rely on sqlx directly?

dynco-nym replied:
see [other reply](https://github.com/nymtech/nym/pull/5673#discussion_r2044546718)

jstuczyn commented on nym-api/build.rs:
why are we changing the build.rs to begin with?

dynco-nym replied:
this generates a git-ignored bash file that allows you to easily shell into a running sqlite DB of nym-api

It's in the `build.rs` because sqlite location changes, depending on target dir which on debug builds have a new name every time you `cargo clean` it

changed: 
- only happens on unix and **not** in CI
- even if function fails, it won't unwrap the whole script

it's very useful while debugging so I wanted to share it. If you feel strongly against, I'll remove it

jstuczyn replied:
I'm not against it - I was just confused by it. if you feel it's useful for debugging, I don't see any reasons to remove it as long as it doesn't fail the build : )

jstuczyn commented on nym-api/src/unstable_routes/models.rs:
```diff
@@ -0,0 +1,57 @@
+    #[schema(value_type = String)]
+    pub delegated: Coin,
+    pub height: u64,
+    #[schema(value_type = String)]
```
`value_type` is not a string, because a `None` will get serialised to `null` rather than `""`. either set it to `Option<String>` or remove the field completely. alternatively change it to `boolean` to indicate whether it's a legacy vested delegation that somebody didn't bother to migrate (and that we should display a big ❗ next to)

dynco-nym replied:
used newtype pattern to correctly derive `ToSchema` for Coin and its fields

jstuczyn commented on nym-api/src/unstable_routes/models.rs:
```diff
@@ -0,0 +1,57 @@
+pub struct NyxAccountDelegationDetails {
+    pub node_id: NodeId,
+    #[schema(value_type = String)]
+    pub delegated: Coin,
```
similarly `Coin` is not a `String` either

jstuczyn replied:
(this comment applies to all following occurences of it)

dynco-nym replied:
fixed

jstuczyn commented on nym-api/src/unstable_routes/models.rs:
```diff
@@ -0,0 +1,57 @@
+// Copyright 2024 - Nym Technologies SA <contact@nymtech.net>
```
nit: 2025

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+// Copyright 2024 - Nym Technologies SA <contact@nymtech.net>
```
nit: 2025

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+#[derive(Clone, Debug, Serialize, Deserialize, ToSchema, utoipa::IntoParams)]
+pub struct AddressQueryParam {
+    #[serde(default)]
+    pub address: String,
```
could we perhaps change it to `AccountId` so that value validation would occur during deserialisation?

dynco-nym replied:
unfortunately not, because `AccountId` is a foreign type which doesn't implement `utoipa::ToSchema`

I could do the whole "newtype pattern with `#[schema(value_type = AccountIdSchema)]`" thing but I don't see an advantage over current approach in that case

jstuczyn replied:
but for the purposes of schema, it's just a string, so couldn't we just put `#[schema(value_type = String)]` there instead like we're doing already with so many types?

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    let mut total_value = 0u128;
+
+    // 1. get balances of chain tokens
+    let balances = state_client
```
if we're only concerned about `unym`, query for only `unym` balance rather than getting `all_balances`. this way you'd also avoid having to mess with `Vec` and adding up amounts to `total_value`

dynco-nym replied:
done

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    })?;
+
+    let node_cache = state.nym_contract_cache();
+    let nym_nodes = node_cache
```
1. first query for account balance before attempting to retrieve and clone internal caches. if account doesn't exist, there's no point in doing any other operation

jstuczyn replied:
2. do we really need to clone the full cache of all nym-nodes? there's quite a lot of them and the cost overhead might not be negligible. perhaps `all_cached_nym_nodes` would have worked better? also do note it does not include **all** nodes. if somebody has an ancient delegation towards a legacy mixnode, it wouldn't come up here

dynco-nym replied:
replaced with `all_cached_nym_nodes`

it should be read from the same list, except this doesn't clone the whole vec?

jstuczyn replied:
yup. exactly

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+                        NyxAccountDelegationRewardDetails {
+                            node_id: delegation.node_id,
+                            rewards: decimal_to_coin(delegation_reward, "unym"),
+                            amount_staked: nym_node_details.original_pledge().clone(),
```
just so we're clear, this is just pledge of the operator, not the full stake

dynco-nym replied:
removing this field

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    // 4. make the map of rewards into a vec and sum the rewards and delegations
+    let accumulated_rewards: Vec<NyxAccountDelegationRewardDetails> =
```
I don't understand. why are you creating the map only to immediately convert it into vec?

dynco-nym replied:
copy/paste from old explorer code, fixed

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    // 5. get vesting account details (if any): none because everyone has already fully vested
+    let vesting_account: Option<NymVestingAccount> = None;
```
well.........
yes. they have fully vested, but not everyone has actually migrated the tokens to liquid. (and we didn't have  time to purge it ourselves)

dynco-nym replied:
as agreed, ignoring vesting info but leaving here in case it's required in the future

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    // 5. get vesting account details (if any): none because everyone has already fully vested
+    let vesting_account: Option<NymVestingAccount> = None;
+
+    if let Some(vesting_account) = vesting_account.clone() {
```
why the clone

dynco-nym replied:
fixed

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    // 6. get operator rewards: is this an operator?
+    let mut operator_rewards = 0;
+
+    for (_, node_details) in nym_nodes.iter() {
```
```suggestion
    for node_details in nym_nodes.values() {
```

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+            operator_rewards += pending_operator_reward;
+
+            // add to totals
+            total_value += pending_operator_reward;
```
`break` at this point, because there's no point in continuing to iterate the map as there can only be a single bond

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    // 7. convert totals
+    let claimable_rewards = Coin::new(claimable_rewards, "unym");
```
I don't really like the hardcoded `unym` everywhere. what if some testnet uses different denom? can you just pull it from one of many `Coin`s you got it at the start?

dynco-nym replied:
good suggestion, reading it from `NymNetworkDetails`

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,207 @@
+    params(AddressQueryParam)
+)]
+#[instrument(level = "info", skip_all, fields(address=address))]
+async fn address(
```
this requires some `AccountCache` behind it, like we have for `ChainStatusCache`. this is a huge potential source of DoS because there's a lot of expensive queries and operations for one account

dynco-nym replied:
[added](https://github.com/nymtech/nym/pull/5673#discussion_r2060104264)

dynco-nym committed:
Fix build.rs build on Windows

dynco-nym committed:
Addressing PR feedback
- not cloning nym nodes from cache
- reduced number of nym nodes kept in memory
- reduced number of iterations to read all data
- removed some fields

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,199 @@
+
+    // 1. get balances of chain tokens
+    let balance = state_client
+        .get_address_balance(&account_id, base_denom)
```
I actually don't remember from the top of my head if, when account doesnt exist, whether it will return an `Err` or `Ok(None)` - can you double check it with some dummy account?

dynco-nym replied:
what do you mean by account not existing?

if the address is invalid, it will fail validation before reaching this step

jstuczyn replied:
by not existing, I mean it not having on-chain representation. there's a difference between 0 balance and no balance (i.e. not existing). for example if you generate a brand new mnemonic, **without sending any transactions**, that account wouldn't exist and if, for example, you queried for its sequence number, it wouldn't return anything

dynco-nym replied:
tested with a fresh address and it handles it fine, filling all fields and returning 200
```
n19ha49y8rwyz2xemf92q7kre80elahr2ykrexm4
```

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,199 @@
+        .await
+        .ok_or_else(AxumErrorResponse::service_unavailable)?
+        .iter()
+        .filter_map(|node_details| {
```
I thought I already left a comment about it - there's no point in iterating the whole cache, just do a `.find()` because there can be at most one node that belongs to particular address

dynco-nym replied:
as you [noticed](https://github.com/nymtech/nym/pull/5673#discussion_r2044979383), I have to go through nodes at least once

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,199 @@
+            if account_id.to_string() == node_details.bond_information.owner.as_str() {
+                let pending_operator_reward = node_details.pending_operator_reward().amount.u128();
+
+                // add operator rewards
```
I think that's something to discuss against requirements. do we want here tokens that could be claimed (i.e. pending rewards) or tokens that are actually associated with the account (`.operator` field - which includes original pledge **and** pending rewards)

jstuczyn replied:
(sorry, I meant to add this comment for `total_value += pending_operator_reward;`)

dynco-nym replied:
currently, unclaimed rewards _are_ displayed in totals so this keeps that behaviour

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,199 @@
+    for delegation in og_delegations.iter() {
+        let node_id = &delegation.node_id;
+
+        if let Some(nym_node_details) = nym_nodes.get(node_id) {
```
ah. right. that's why you need all nym nodes in a map

jstuczyn commented on nym-api/src/unstable_routes/account.rs:
```diff
@@ -0,0 +1,199 @@
+    params(AddressQueryParam)
+)]
+#[instrument(level = "info", skip_all, fields(address=address))]
+async fn address(
```
if possible, can you perhaps split this function into some smaller chunks, say one for retrieving delegation rewards, one for operator stuff, etc. i think it would make it slightly easier to read

dynco-nym replied:
done

dynco-nym committed:
Fix total_delegations

dynco-nym committed:
Optimize nym_nodes hashmap

dynco-nym committed:
Split flow into functions

dynco-nym committed:
Remove vesting info

dynco-nym committed:
Add caching for endpoint

dynco-nym commented on nym-api/src/unstable_routes/account/cache.rs:
```diff
@@ -0,0 +1,115 @@
+    pub(crate) fn new() -> Self {
+        // TODO dz values should come from config
+        let cache_ttl = Duration::from_secs(60);
+        let max_capacity = 1000;
```
@jstuczyn sensible defaults?

dynco-nym committed:
Cache optimizations

jstuczyn commented on nym-api/src/support/http/state.rs:
```diff
@@ -292,4 +295,12 @@ impl AppState {
+        self,
+        account_id: nym_validator_client::nyxd::AccountId,
+    ) -> Result<NyxAccountDetails, AxumErrorResponse> {
+        let address_info_cache = self.address_info_cache.clone();
```
I'm not sure I understand the logic here, we're cloning the address cache and consuming `self`, only to call a method on address cache with `self` as the argument, i.e. changing `state.foo(cache)` to `cache.bar(state)`. why not just do everything here without that clone?

dynco-nym replied:
fixed

jstuczyn commented on nym-api/src/unstable_routes/account/data_collector.rs:
```diff
@@ -0,0 +1,214 @@
+            .nyxd_client
+            .get_address_balance(&self.account_id, &self.base_denom)
+            .await?
+            .unwrap_or_else(|| nym_validator_client::nyxd::Coin::new(0u128, &self.base_denom));
```
so what was the result of your test, i.e. if the account didn't exist did it return an `Err` or `Ok(None)`?
essentially, would this function return an error so no further queries would be invoked?

dynco-nym replied:
as we discussed in private: this returns `Ok(Some)` in all cases

there is [get_account](https://github.com/nymtech/nym/blob/develop/common/client-libs/validator-client/src/nyxd/cosmwasm_client/client_traits/query_client.rs#L90), which fails with `Err - not found`

but that's an overkill for this. To reduce number of queries, we fetch balance with `get_address_balance`, assume account with 0 unym is freshly created, has no delegtions and isn't an operator and we return early in that case

jstuczyn reviewed (approved):
i think I added few nits and questions, but lgtm

dynco-nym committed:
Return early if balance is 0

dynco-nym committed:
Refactor state cloning shenanigans
